doc/dev/static_type_checking_cheat_sheet.md

```diff
@@ -0,0 +1,410 @@
+        ...
+
+# No: the function calls len(), Iterable doesn't support it
+def create_batch(entities: Iterable[Entity]) -> None:
+    if len(entities) > 1:
+        ...
+```
+
+- If you don't need to mutate the collection, prefer parameter types `Sequence` over `MutableSequence`, `Mapping` over `MutableMapping`, etc.
+- Consider using `typing.TypedDict` if a dictionary has a fixed set of keys. This is especially useful if a user needs to construct the dict. Do make this importable from the public namespace.
+- Consider using `typing.NamedTuple` if your tuple should have named fields. Do make this importable from the public namespace.
+
+
+### Overloads
+
+- Use `typing.overload` if your function takes different combinations of typed arguments and the input arguments might inform the return type.
+
+```python
+from typing import overload, Union
+
+
+@overload
+def analyze(text: str, task: LanguageDetection) -> LanguageDetectionResult:
+    ...
+
+
+@overload
+def analyze(text: str, task: EntityRecognition) -> EntityRecognitionResult:
+    ...
+
+
+@overload
+def analyze(text: str, task: SentimentAnalysis) -> SentimentResult:
+    ...
+
+# actual implementation
+def analyze(text: str, task: Union[SentimentAnalysis, EntityRecognition, LanguageDetection]) -> Union[
+    SentimentResult, EntityRecognitionResult, LanguageDetectionResult]:
+    return _analyze(text, task)
+```
+
+- You should add type hints to the actual implementation for overloads. This helps with our internal tooling and ensures the implementation is type checked.
+
+### Variadic arguments
+
+- Do include type hints for variadic arguments, like `*args` and `**kwargs`.
+
+```python
+from typing import Any
+
+# args seen as tuple[str] by type checker
+# kwargs seen as dict[str, Any] by type checker
+def begin_operation(*args: str, **kwargs: Any) -> None:
+    ...
+```
+
+### Structural subtyping / Protocols
+
+- Do support the "duck typing" of Python with `typing.Protocol`.
+- Mark your `Protocol`s as `@runtime_checkable` so users can use them in `isinstance` and `issubclass` checks.
+
+```python
+from typing_extensions import Protocol, runtime_checkable
+
+@runtime_checkable
+class TokenCredential(Protocol):
+    """Protocol for classes able to provide OAuth tokens."""
+
+    def get_token(
+        self, *scopes: str, claims: Optional[str] = None, tenant_id: Optional[str] = None, **kwargs: Any
+    ) -> AccessToken:
+        ...
+```
+
+### Generics
+
+- Use `typing.TypeVar` to let functions/classes work with arguments of any type and retain relationships between arguments and their return values.
+
+```python
+from typing import TypeVar, Sequence, Any
+
+# No:
+def choose(items: Sequence[Any]) -> Any:
+    ...
+
+# Yes:
+T = TypeVar("T")
+
+def choose(items: Sequence[T]) -> T:
+    ...
+```
+
+- Try to constrain your `TypeVar`'s by specifying types or using the `bound` keyword argument.
+
+```python
+from typing import TypeVar
+
+T = TypeVar("T", int, str)  # limited to int or str
```

Comment:
I would honestly not even mention this form here. It has some unexpected behaviors and you should be able to specify a union upper bound to get what most users expect from such a function. (If you decide to delete this, the bullet text on line 266 should also be updated.)